README.md

```diff
@@ -14,11 +16,127 @@ device drivers, even for protcols that are naturally packet-based such as
+
+- infrastructure to convert byte-oriented streams into streams of validated
+  packages
+- library-level test harness
+- generic oroGen integration (for Rock users)
+- built-in support for multiple type of I/O:
+  - serial
+  - udp server: listens on a UDP port. Sends to the IP of the last received
+    UDP message.
+  - udp client: connects to a given IP and port.
+  - tcp client: connects to a given IP and port.
+  - file-based (e.g. for named pipes or Unix sockets)
+
+Note that `iodrivers_base` can (and maybe should) be used even for
+datagram-oriented mediums such as UDP. These mediums do not need
+the packet-reassembly, but do need packet validation.
+
+## Writing a device driver using `iodrivers_base::Driver`
+
+One starts by subclassing `iodrivers_base::Driver` and providing the size of
+the driver's static internal buffer. This buffer should be a multiple of the maximum
+packet size (twice as big is usually enough):
+
+~~~cpp
+class Driver : iodrivers_base::Driver
+{
+    static const int MAX_PACKET_SIZE = 32;
+    static const int INTERNAL_BUFFER_SIZE = MAX_PACKET_SIZE * 4;
+public:
+    Driver();
+
+};
+~~~
+
+~~~cpp
+Driver::Driver()
+    : iodrivers_base::Driver::Driver(INTERNAL_BUFFER_SIZE);
+{
+}
+~~~
+
+The second required step is to implement `iodrivers_base::Driver::extractPacket`.
+This is the method that tells `iodrivers_base`'s internal logic what is meaningful
+data and what isn't.
+
+See below the pseudo-code implementation of most `extractPacket` methods. Note
+that the amount of validation you do in `extractPacket` may vary, but usually
+"more is better", as it is validation that won't have to be done later.
+The return value is fully documented within the method documentation. Check it
+out for a complete specification.
+
+~~~cpp
+int extractPacket(uint8_t const* buffer, int buffer_size) {
+    if (not enough bytes in buffer to contain a start code) {
+        return 0; // wait for new bytes
+    }
+
+    if (!buffer starts with packet start code) {
+        for (int i = 1; i < buffer_size; ++i) {
+            if (packet start code found at i) {
+                return -i; // skip i bytes from buffer and call again
+            }
+        }
+        return -buffer_size; // discard the whole buffer
+    }
+
+    if (not enough bytes in buffer to validate packet starting at zero) {
+        return 0; // wait for new bytes
+    }
+    else if (packet starting at zero is valid) {
+        return packet_size;
+    }
+    else {
+        return -1; // discard first byte and start searching again
+    }
+}
+~~~
+
+## Supported URIs
+
+- serial ports are open with `serial:/${DEVICE_PATH}[:BAUDRATE]`, e.g.
+  `serial:///dev/ttyUSB0:115200`
+- a UDP server is open with `udpserver://[PORT]`. The UDP server will
+  receive from any source, and send to the IP/Port of the last received
+  packet it received
+- a UDP client is open with `udp://host:PORT`. The UDP client will
```

Comment:
I would document this one as `udp://host:outport[:inport]` instead

Reply:
used `:remote_port[:local_port]`, I feel it easier to understand